File: prover/checker.py
from prover.datatypes import *
from prover.prelude import *
from prover.errors import *

from functools import partial
from typing import List, Dict

def subst(name : Name, μ : Term, τ : Term) -> Term:
    if isinstance(τ, Lit):
        return τ
    elif isinstance(τ, Var):
        if τ.name == name:
            return μ
        else:
            return τ
    elif isinstance(τ, Symtree):
        return Symtree(maplist(partial(subst, name, μ), τ.children))
    elif isinstance(τ, Hole):
        return τ

def multisubst(substitutions : Dict[Name, Term], τ : Term) -> Term:
    salt = gensym()
    for name, _ in substitutions.items():
        τ = subst(name, Var(name + salt), τ)

    for name, μ in substitutions.items():
        τ = subst(name + salt, μ, τ)

    for name, _ in substitutions.items():
        τ = subst(name + salt, Var(name), τ)
    return τ

def prune(substs : Dict[Name, Term], τ : Term):
    if isinstance(τ, Var):
        if τ.name in substs:
            return substs[τ.name]
        else:
            return τ
    else:
        return τ

def unify(substs : Dict[Name, Term], α : Term, β : Term) -> bool:
    φ, ψ = prune(substs, α), prune(substs, β)
    if isinstance(φ, Var):
        substs[φ.name] = ψ
        return True
    elif isinstance(φ, Lit) and isinstance(ψ, Lit):
        return φ.name == ψ.name
    elif isinstance(φ, Symtree) and isinstance(ψ, Symtree):
        if len(φ.children) != len(ψ.children):
            return False
        for δφ, δψ in zip(φ.children, ψ.children):
            if not unify(substs, δφ, δψ):
                return False
        return True
    elif isinstance(φ, Hole):
        return True
    else:
        return False

def occurs(τ : Term, name : Name):
    if isinstance(τ, Var):
        return τ.name == name
    elif isinstance(τ, Lit):
        return False
    elif isinstance(τ, Symtree):
        for σ in τ.children:
            if occurs(σ, name):
                return True
        return False
    else:
        return False

def even(φ : Term, ψ : Term):
    if φ != ψ:
        raise UnificationError(φ, ψ)

def lookup(ctx : Dict[Name, InferenceRule], name : Name) -> InferenceRule:
    if name in ctx:
        return ctx[name]
    else:
        raise NotDefinedError(name)

def getbound(bound : List[Term], τ : Term) -> List[Name]:
    vars = []

    for formula in bound:
        substs = {}
        if unify(substs, formula, τ):
            for name, σ in substs.items():
                if not isinstance(σ, Var):
                    raise VerificationError("“%s” expected to be a variable" % σ)
            vars = maplist(lambda σ: σ.name, substs.values())
            break

    if isinstance(τ, Symtree):
        for child in τ.children:
            vars.extend(getbound(bound, child))
    return vars

def checksubst(bound : List[Term], substitutions : Dict[Name, Term], τ : Term):
    BV = getbound(bound, τ)
    for name, σ in substitutions.items():
        if name in BV and not isinstance(σ, Var):
            raise VerificationError(
                "cannot replace bound variable “{0}” with a constant “{1}”".format(
                    name, σ
                )
            )
        elif isinstance(σ, Var) and σ.name in BV:
            raise VerificationError(
                "cannot replace “{0}” with “{1}”, because “{1}” is bound".format(
                    name, σ.name
                )
            )

def sorry(tree : Sorry, τ : Term):
    print("%s: expected “%s”" % (tree.name, τ))

def infer(ctx : Dict[Name, InferenceRule], bound : List[Term],
          tree : Derivation) -> Term:
    statement = lookup(ctx, tree.edge)
    if len(tree.children) != len(statement.premises):
        raise VerificationError(
           "%s expects %d arguments, but got %d" % (
               tree.edge,
               len(statement.premises),
               len(tree.children)
           )
        )

    for premise, child in zip(statement.premises, tree.children):
        checksubst(bound, tree.substitutions, premise)
        expected = multisubst(tree.substitutions, premise)
        if isinstance(child, Sorry):
            sorry(child, expected)
        elif isinstance(child, Proof):
            τ = infer(ctx, bound, child)
            even(expected, τ)

    checksubst(bound, tree.substitutions, statement.conclusion)
    return multisubst(tree.substitutions, statement.conclusion)

def check(ctx : Dict[Name, InferenceRule], bound : List[Term],
          τ : Term, tree : Derivation):
    # top-level sorry
    if isinstance(tree, Sorry):
        sorry(tree, τ)
    else:
        π = infer(ctx, bound, tree)
        even(π, τ)

File: checker.ml
open Datatypes
open Prelude
open Errors

let rec index idx = function
  | Lit x       -> Lit x
  | FVar (x, _) -> FVar (x, -1)
  | Var (x, _)  -> Var (x, idx)
  | Symtree xs  -> Symtree (List.map (index idx) xs)
  | Hole        -> Hole

let rec unify substs patt term =
  let err = MatchError (patt, term) in

  let omega = prune substs patt in
  let tau   = prune substs term in

  match omega, tau with
  | FVar a, FVar b when a = b -> substs
  | Var a, Var b when a = b -> substs
  | Var name, _ -> Sub.add name tau substs
  | _, Var name -> Sub.add name omega substs
  | Lit a, Lit b -> if a = b then substs else raise err
  | Symtree xs, Symtree ys ->
    if List.length xs <> List.length ys then raise err
    else List.fold_left2 unify substs xs ys
  | Hole, _ -> substs
  | _, _ -> raise err

let rec occurs x : term -> bool = function
  | FVar y | Var y -> x = y
  | Symtree xs     -> List.exists (occurs x) xs
  | _              -> false

let even fi tau =
  if fi <> tau then raise (UnificationError (fi, tau))
  else ()

let lookup ctx name =
  match Env.find_opt name ctx with
  | Some v -> v
  | _      -> raise (NotDefinedError name)

let rec getBound bound tau =
  let formula = findMap (fun x -> getMatch Sub.empty x tau) bound in
  let vars =
    begin match formula with
      | Some substs ->
        List.map
          (function
          | _, FVar name
          | _, Var name -> name
          | _, omega    -> raise (ExpectedVariable omega))
          (Sub.bindings substs)
      | None -> []
    end in
  match tau with
  | Symtree xs -> vars @ List.concat (List.map (getBound bound) xs)
  | _          -> vars

let checkSubst bound substs tau =
  let bvars = ref (getBound bound tau) in
  Sub.iter (fun name omega ->
    match omega with
    | FVar name' | Var name' when occurs name tau ->
      (* Variable cannot be replaced with bound variable *)
      if List.mem name' !bvars then
        ReplacingWithBound (fst name, substs)
        |> raise
      else ();

      (* Bound variable cannot be replaced with present in term variable *)
      if List.mem name !bvars then
        if occurs name' tau then
          ReplacingBoundWith (fst name, fst name')
          |> raise
        else bvars := name' :: !bvars
      else ()
    | _ ->
      (* Bound variable cannot be replaced with a constant *)
      if List.mem name !bvars then
        raise (ReplacingBoundWithConstant (fst name, omega))
      else ())
    substs

let substitute bound substs tau =
  checkSubst bound substs tau;
  multisubst substs tau

let synth ctx bound tau xs =
  let goals : (term list) ref = ref [tau] in
  let rwr   : sub ref = ref Sub.empty in

  List.iteri (fun idx (name, substs) ->
    let rule = lookup ctx name in
    let conclusion = index idx rule.conclusion in
    let premises = List.map (index idx) rule.premises in
    Sub.iter (fun (name, _) tau -> rwr := Sub.add (name, idx) tau !rwr) substs;

    let goal = pop goals in
    let expected = substitute bound !rwr conclusion in
    let unifying = unify Sub.empty expected goal in
    Sub.iter (fun k v -> rwr := Sub.add k v !rwr) unifying;
    rwr := Sub.map (substitute bound !rwr) !rwr;

    goals := premises @ !goals;
    goals := List.map (substitute bound !rwr) !goals) xs;

  if List.length !goals > 0 then
    raise (Goals !goals)
  else !rwr

let check ctx bound tau xs =
  let goals : (term list) ref = ref [tau] in
  let rwr = synth ctx bound tau xs in

  List.iteri (fun idx (name, _) ->
    let rule       = lookup ctx name in
    let conclusion = substitute bound rwr (index idx rule.conclusion) in
    let premises   = List.map (index idx >> substitute bound rwr) rule.premises in

    let goal = pop goals in
    even goal conclusion;

    goals := premises @ !goals) xs